# Generated by sila2.code_generator; sila2.__version__: 0.8.0
from __future__ import annotations

from typing import Set

from sila2.client import SilaClient
from sila2.framework import FullyQualifiedFeatureIdentifier

from .axispositioncontroller import AxisPositionControllerClient, AxisPositionControllerFeature, InvalidAxisIdentifier
from .axissystemcontrolservice import AxisSystemControlServiceClient
from .axissystempositioncontroller import (
    AxisSystemPositionControllerClient,
    AxisSystemPositionControllerFeature,
    MovementBlocked,
)


class Client(SilaClient):

    AxisPositionController: AxisPositionControllerClient

    AxisSystemPositionController: AxisSystemPositionControllerClient

    AxisSystemControlService: AxisSystemControlServiceClient

    _expected_features: Set[FullyQualifiedFeatureIdentifier] = {
        FullyQualifiedFeatureIdentifier("org.silastandard/core/SiLAService/v1"),
        FullyQualifiedFeatureIdentifier("de.cetoni/motioncontrol.axis/AxisPositionController/v1"),
        FullyQualifiedFeatureIdentifier("de.cetoni/motioncontrol.axis/AxisSystemPositionController/v1"),
        FullyQualifiedFeatureIdentifier("de.cetoni/motioncontrol.axis/AxisSystemControlService/v1"),
    }

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._register_defined_execution_error_class(
            AxisPositionControllerFeature.defined_execution_errors["InvalidAxisIdentifier"], InvalidAxisIdentifier
        )

        self._register_defined_execution_error_class(
            AxisSystemPositionControllerFeature.defined_execution_errors["MovementBlocked"], MovementBlocked
        )
